File: flow-client/src/main/java/com/vaadin/client/communication/PollConfigurator.java
/*
 * Copyright (C) 2000-2024 Vaadin Ltd
 *
 * This program is available under Vaadin Commercial License and Service Terms.
 *
 * See <https://vaadin.com/commercial-license-and-service-terms> for the full
 * license.
 */
package com.vaadin.client.communication;

import com.vaadin.client.flow.StateNode;
import com.vaadin.client.flow.nodefeature.MapProperty;
import com.vaadin.client.flow.nodefeature.NodeMap;
import com.vaadin.flow.internal.nodefeature.NodeFeatures;
import com.vaadin.flow.internal.nodefeature.PollConfigurationMap;

/**
 * Observes the poll configuration stored in the given node and configures
 * polling accordingly.
 *
 * @author Vaadin Ltd
 * @since 1.0
 */
public class PollConfigurator {

    private PollConfigurator() {
        // No instance should ever be created
    }

    /**
     * Observes the given node for poll configuration changes and configures the
     * given poller accordingly.
     *
     * @param node
     *            the node containing the poll configuration
     * @param poller
     *            the poller to configure
     */
    public static void observe(StateNode node, Poller poller) {
        NodeMap configurationMap = node.getMap(NodeFeatures.POLL_CONFIGURATION);
        MapProperty pollIntervalProperty = configurationMap
                .getProperty(PollConfigurationMap.POLL_INTERVAL_KEY);
        pollIntervalProperty.addChangeListener(e -> {
            int interval = (int) (double) e.getNewValue();
            poller.setInterval(interval);
        });
    }

}

File: flow-client/src/main/frontend/internal/client/communication/PollConfigurator.ts
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

// TypeScript port of com.vaadin.client.communication.PollConfigurator, built
// alongside the Java version. It observes the poll-interval property in the
// node's POLL_CONFIGURATION feature and reconfigures the poller on change.

import type { StateNode } from '../flow/StateNode';
import type { Poller } from './Poller';
import { NodeFeatures } from '../../flow/internal/nodefeature/NodeFeatures';

// com.vaadin.flow.internal.nodefeature.PollConfigurationMap.POLL_INTERVAL_KEY
const POLL_INTERVAL_KEY = 'pollInterval';

// The poller is named by the method it is driven through rather than as a whole
// Poller, so a suite can observe the configured interval directly; the type is
// still derived from the ported class, so it cannot drift from it.
type ConfigurablePoller = Pick<Poller, 'setInterval'>;

/**
 * Observes the node's poll configuration and configures the poller on change.
 * Mirrors PollConfigurator.observe.
 */
export function observe(node: StateNode, poller: ConfigurablePoller): void {
  const pollIntervalProperty = node.getMap(NodeFeatures.POLL_CONFIGURATION).getProperty(POLL_INTERVAL_KEY);
  pollIntervalProperty.addChangeListener((event) => {
    poller.setInterval(Math.trunc(Number(event.getNewValue())));
  });
}
